File: 백준/Platinum/11400. 단절선/단절선.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Main {
    static StringTokenizer st;
    static List<Integer>[] edges;
    static List<Edge> res;
    static int[] ids;
    static int V, E, id = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        V = nextInt();
        E = nextInt();

        ids = new int[V + 1];
        edges = new List[V + 1];

        for (int u = 1; u <= V; u++) {
            edges[u] = new ArrayList<>();
        }

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int u = nextInt();
            int v = nextInt();

            edges[u].add(v);
            edges[v].add(u);
        }

        res = new ArrayList<>();

        travel(1, 1);

        StringBuilder sb = new StringBuilder();

        res.sort((a, b) -> {
            if (a.u == b.u) return a.v - b.v;

            return a.u - b.u;
        });

        sb.append(res.size()).append('\n');

        for (Edge edge : res) {
            sb.append(edge.u).append(' ').append(edge.v).append('\n');
        }

        System.out.println(sb);
        br.close();
    }

    static int travel(int p, int u) {
        ids[u] = ++id;

        int minId = id;

        for (int v : edges[u]) {
            if (v == p) continue;

            int cid;

            if (ids[v] > 0) {
                cid = ids[v];
            } else {
                cid = travel(u, v);
            }

            if (ids[u] < cid) {
                res.add(new Edge(u, v));
            }

            minId = Math.min(minId, cid);
        }

        ids[u] = minId;

        return minId;
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }

    static class Edge {
        int u;
        int v;

        Edge(int u, int v) {
            if (u < v) {
                this.u = u;
                this.v = v;
            } else {
                this.u = v;
                this.v = u;
            }
        }
    }
}

File: 백준/Platinum/11400. 단절선/단절선.kt
import java.io.StreamTokenizer

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    fun readInt() = run {
        nextToken()
        nval.toInt()
    }

    val n = readInt()
    val k = readInt()
    val graph = Graph(n)

    repeat(k) {
        graph.addEdge(readInt(), readInt())
    }

    val res = graph.findSpfEdges()

    with(System.out.bufferedWriter()) {
        append("${res.size}\n")
        res.forEach { append("${it.u} ${it.v}\n") }
        flush()
        close()
    }
}

class Graph(val n: Int) {
    val edges = Array(n + 1) { ArrayList<Int>() }
    val spfEdges = ArrayList<Point>()
    val ids = IntArray(n + 1)
    var id = 0

    fun addEdge(u: Int, v: Int) {
        edges[u] += v
        edges[v] += u
    }

    fun findSpfEdges(): ArrayList<Point> {
        for (u in 1..n) {
            if (ids[u] == 0) travel(u, 0)
        }

        spfEdges.sortWith(compareBy(Point::u, Point::v))

        return spfEdges
    }

    fun travel(u: Int, p: Int): Int {
        ids[u] = ++id

        var minId = id

        for (v in edges[u]) {
            if (v == p) continue

            if (ids[v] > 0) {
                minId = minOf(minId, ids[v])
                continue
            }

            val next = travel(v, u)

            if (next > ids[u]) {
                spfEdges += Point(u, v)
            }

            minId = minOf(minId, next)
        }

        return minId
    }
}

class Point(var u: Int, var v: Int) {
    init {
        if (u > v) {
            val temp = u

            u = v
            v = temp
        }
    }
}